### ISSUE TITLE
Pass to or make GalSim call for realistic pupil.

### ISSUE THREAD

**Issue by cwwalter** [LSSTDESC/imSim#28]
So, I don't forget:

We need to generate a realistic aperture with spider etc.  As mentioned in #16 we may also want to make sure vignetting can be handled by correct pupil obscuration.

I recall @rmandelb was already working on a realistic LSST geometry in GalSim itself.  So perhaps we can just call that? If not, and we need to implement it, we either need to do it there or define it here and then pass it to the proper routines if that is possible.

**Comment by rmandelb** [comment]
I was working on that, but got distracted by some roadblock (I was missing some info I needed) and then started teaching and it didn't happen.  Let me try to sort out the roadblock this week, so I can put the software in place during hack week.  Would that work for your schedule? 

And I would like to have this in GalSim rather than in here, for parallelism with what we provide for WFIRST in GalSim.

**Comment by cwwalter** [comment]
That sounds great Rachel.  If there is an issue for this in the GalSim repo for this could you also reference it in this issue so we know when we can work on calling it?

You might also want to make a new GalSim issue (so as not to forget) related to Josh's comment in #11 about vignetting.

**Comment by rmandelb** [comment]
Issue in GalSim repo: https://github.com/GalSim-developers/GalSim/issues/556
(kind of generic one but you can see the discussion of the spider at the end; I was waiting on something from Aaron that I will chase down this week)

Vignetting is part of one GalSim issue ( https://github.com/GalSim-developers/GalSim/issues/553 ) already; is that enough to address Josh's comment?  (I don't see a comment from him on there so I might not be looking in the right place)

**Comment by cwwalter** [comment]
> Vignetting is part of one GalSim issue ( GalSim-developers/GalSim#553 ) already; is that enough to address Josh's comment? (I don't see a comment from him on there so I might not be looking in the right place)

Sorry.. I typed the wrong issue... it should have been #16.  Here is he comment:

"Pupil obscuration should work for getting the vignetted PSF right, but it won't get the amplitude right in the current GalSim implementation. (Right now the GalSim optics PSF is always normalized to unit integrated surface brightness). I think it would be relatively straightforward to do the right thing though and scale the PSF amplitude by the area of the unobscured pupil."

I think this is a bit different than GalSim-developers/GalSim#553 which would just be applying vignetting to the produced image.

**Comment by cwwalter** [comment]
@rmandelb What was the status of the work in GalSim-developers/GalSim#556?  Is that on hold until performance issues are worked out?   I don't think we need this for DC2 but if it was already there we could turn it on.

**Comment by rmandelb** [comment]
@cwwalter - yes, unfortunately we didn't sort out the performance issues; I only got as far as demonstrating that most of my ideas for speeding it up don't work by the end of hack week.  However, I do want to direct your attention to this comment: 
https://github.com/GalSim-developers/GalSim/issues/556#issuecomment-258957810

It shows a high-res image of the optical PSF with obscuration and without aberrations for two cases:
- using the standard GalSim implementation of struts
- using the proper LSST spider

They are really very similar, and the former is 40x faster.  Given this plus the fact that the atmospheric PSF likely makes these details almost completely irrelevant, I think it would be completely reasonable for you to include the optical PSF with obscuration + 4 struts using the default settings rather than waiting on the more exact pupil plane configuration.  I can provide the appropriate kwargs for this if helpful.

**Comment by cwwalter** [comment]
@rmandelb OK great. 

 I agree that the default option would be  fine for what we are doing now. So, it looks like we would do this with galsim.phase_psf.Aperture.__init__ by setting "nstruts" correct?

@danielsf, can confirm but right now we are either using one of 

"__all__ = ["PSFbase", "DoubleGaussianPSF", "SNRdocumentPSF",
11	           "Kolmogorov_and_Gaussian_PSF"]
12	"

in https://github.com/lsst/sims_GalSimInterface/blob/ab2aa51b504e5a3ef0531f4b4f1261646ee439ca/python/lsst/sims/GalSimInterface/galSimPSF.py

As @dkirkby pointed out in his presentation these are also supposed to include the instrument response.  So, we should decide whether we want these diffraction spikes from the spiders in DC2 and then, if so, consider how to include them.

**Comment by rmandelb** [comment]
I normally do it with galsim.OpticalPSF (and can provide you with the code to do so), but yes, ```nstruts``` is what you want.

Yes, as David said, things like Kolmogorov+Gaussian or double Gaussians are supposed to include the instrument response.  For this purpose I would recommend making a new option that you could call ```Kolmogorov_and_OpticalPSF``` or something like that, which involves convolving a Kolmogorov and an OpticalPSF object.  If helpful I can provide code for that.

But one final question: what are your requirements on timing for image rendering?  Using the approximate strut configuration is 40x faster than using the spider, but even the approximate strut configuration is slower than Kolmogorov+Gaussian.  Would you like a timing estimate for a typical Kolmogorov and an LSST-like approximate strut configuration so you can decide whether this is going to be a problem?  I can dig up my scripts that gave the timing estimates on that GalSim branch quite easily and provide you with those numbers - unless some other step in the simulation process is already the tall pole?

**Comment by danielsf** [comment]
Chris' characterization of what we are currently doing is correct.

**Comment by cwwalter** [comment]
I think what I would like is for DC1 and other tests as we are doing now to keep that options that we have. For DC2+ and "realistic" simulations I would like to pair this with Josh's atmospheric ray tracing code to produce a realistic PSF.  Does this seem reasonable to people?

I suspect that the raytracing will dominate the processing time.  But, I don't know.   Once that is working (there is currently no chromatic response there) we should try to profile these various parts including the strut configuration for the optical PSF.   If you do already have some #s that tell us how much time including the struts adds that would be great if you could share them here.

**Comment by rmandelb** [comment]
> I think what I would like is for DC1 and other tests as we are doing now to keep that options 
> that we have.

Absolutely agree; I definitely was not proposing any changes for any sims being made now or in the very near future.

> For DC2+ and "realistic" simulations I would like to pair this with Josh's atmospheric ray 
> tracing code to produce a realistic PSF.

I'm worried about issues of speed.  Hopefully @jmeyers314 can comment on this.

For what it's worth, I do also think this corresponds to ratcheting up the complexity by several levels.  Do we want to do that all at once?  The progression as I see it would be something like:

- a simple model that is meant to include the effect of optical + atmospheric PSFs, like double Gaussian or Moffats  -> that's what you're doing now
- a Kolmogorov PSF with some power spectrum for the shape correlation function convolved with an optical PSF model that has some model for the aberrations -> next step up in complexity
- full ray tracing of stochastic atmosphere plus optical PSF -> highest level of complexity

So in my head, these 3 steps would be something like DC1, DC2, DC3.  Is there a reason to not step up gradually?  I could definitely imagine some: for example, if we're really worried about having fully realistic atmospheric effects in DC2 to test the DM PSF modeling routines that are going to be rolled out at some point in the coming ~year (? right @jmeyers314 ?), so that we can provide them with feedback on algorithmic performance before DC3, that would be a good reason to jump straight from the lowest to the highest level of realism.

**Comment by cwwalter** [comment]
So, my thinking on this is that we want to make sure we have the basic versions of major pieces we want to use eventually in DC3 in place by DC2. This is because DC3 will be a really big step up (at least a factor of 10 in size) and we want the opportunity to really stress test these features first in DC2.  Also, we need to be able to start testing some of the PSF modeling and sensor effect correction routines in what (in the end will) will likely be the next two years (if you think about now to when DC3 would be ready).

For this reason, the three main things I would like to see implemented by DC2 are:

- Sky varying version of the ESO sky model (now the whole sky is set to the pointing position)

- Josh's ray tracing code with 2nd kick and chromatic dependance.   

- Sensor model with saturation, tree rings and BF along with electronics readout.

All of these things are within reach now I believe.

For DC3 I would expect that we would add refinements on top of this, vignetting, more sensor effects, clouds etc. etc.  These three effects are also important for bringing us more in parity with the the PhoSim feature list. If as Rachel says any of these things have serious speed issues we may need to re-evaluate, but we still should find that out now if we can I think.  There are more serious issues (like optical raytracing) that for right now are off the table all together but one could imaging if large progress was made could be considered on the DC3 time scale.

**Comment by connolly** [comment]
I think it is important to be as feature complete as we can be for DC2
(especially the components that may be compute intensive) as the
computational requirements for DC3 are quite significant and it will be
hard to get these resources from DOE. It would be better to know as early
as possible whether we need to dial back the fidelity (and community
expectations) for DC3 or to focus more resources on optimizing the code.

**Comment by cwwalter** [comment]
@rmandelb I was thinking about @jmeyers314 's work and the new aberration work and was wondering how if at all we could deal with diffraction spikes from the spider.  Then I remembered this issue and 

https://github.com/GalSim-developers/GalSim/issues/556

which is now merged.

So that gave us a new optical PSF with spikes (but it was slow) is that correct?

**Comment by jmeyers314** [comment]
It *might* be possible to work diffraction spikes into the second kick.  That would mean calculating the second kick via a simulation and storing it in an InterpolateImage instead of calculating it directly as a numerical integral.  If that works though, then I think you'd only have to do it once per exposure so it may be okay if it's somewhat expensive.

Note though that I think you wouldn't completely capture the right effect this way since at some (probably small) level the projection of the spiders onto the pupil depends on where you are in the field of view.

**Comment by rmandelb** [comment]
Actually @cwwalter , embarrassingly enough, this was never merged because it was slow as molasses via DFT and I really wanted to speed it up.

I don't think it would be hard to update the branch and get it reviewed, and just warn people to never use it in DFT mode, only photon-shooting.  Would that be helpful?

**Comment by cwwalter** [comment]
> I don't think it would be hard to update the branch and get it reviewed, and just warn people to never use it in DFT mode, only photon-shooting. Would that be helpful?

Oh! I somehow tricked myself because at the very end I noticed a "merged" button but in fact that is from a issue that referenced this one.

> I don't think it would be hard to update the branch and get it reviewed, and just warn people to never use it in DFT mode, only photon-shooting. Would that be helpful?

Hmm... I'm actually not sure.  I mean it's probably good to have as a feature and it might also be good to have as a reference for comparison (like what Josh just did for the DFT for ray-traced atmosphere) but for actually getting these effects working with Josh's code I guess we need to do something else like thinking about how to make it work with the 2nd kick as Josh suggested above.

**Comment by cwwalter** [comment]
Work on spiders, diffraction spikes and vignetting will first be addressed though #10 when Josh's Batoid code is integrated with imSim .  Closing this issue now.
